client/app/components/dashboards/EditParameterMappingsDialog.jsx

```diff
@@ -44,7 +59,9 @@ class EditParameterMappingsDialog extends React.Component {
 
     Promise.all(map(widgetsToSave, w => w.save()))
       .then(() => {
-        this.props.dialog.close();
+        const paramValuesSig = this.constructor.getParamValuesSignature(this.state.parameterMappings);
```

Comment:
I was thinking about that `getParamValuesSignature` function, and IMO it's not self-explaining. Look here: before we saved a widget (in the beginning of this function) we already have two snapshots: `widget.options.parameterMappings` and `this.state.parameterMappings`. Let's just create a function that will take two that snapshots and compare them, and call it here. You can detect if any parameter changed (instead of serializing all). WDYT?

Reply:
The was my first attempt.
But it mandates the following steps:
1. Diff objects
2. Omit irrelevant changes - type change, title change

The signature solution is shorter, simpler and more effective imo.

Reply:
No need to diff objects. I mean something like this:
```javascript
function mapingsChanged(prev, next) {
  const s = JSON.stringify;
  const c = v => v.mapTo;
  
  // not sure if `m.value || m.param.value` will work, maybe 
  // need `MappingType.StaticValue` check as in `getParamValuesSignature`
  const prevSnapshot = sortBy(prev, c)
    .map(m => [ m.mapTo, m.value || m.param.value ]);
  const nextSnapshot = sortBy(next, c)
    .map(m => [ m.mapTo, m.value || m.param.value ]);

  return s(prev) !== s(next);
```

Note that I also did two things to make that check as stable as possible: added names of parameters alongside to values, and sorted them by param name.

Reply:
That'll work but now you'll have to store a clone of original `paramterMappings` which was also a consideration of mine. But if you feel this has value, I'll change it.

Reply:
P.S. `m.mapTo` will not work, should use `m.name`

Reply:
@ranbena You already have that snapshot: 
https://github.com/getredash/redash/blob/2c1400d3235950989fcf9937a19e7af9cab051ac/client/app/components/dashboards/EditParameterMappingsDialog.jsx#L38
On the left side of this assignment you have previous mappings, on the right - new ones :slightly_smiling_face:

Reply:
You're sure about this? I checked it and it's not the same output.

Reply:
What do you mean?

Reply:
The object structure of `this.state.parameterMappings`  and `widget.options.parameterMappings` is not the same.

Reply:
`editableMappingsToParameterMappings` returns the same structure as in `widget.options.parameterMappings` (that function exists to create structure that could be assigned to `widget.options.parameterMappings`)

Reply:
You should turn that line into something like this:
```javascript
const prevMappings = widget.options.parameterMappings;
const newMappings = editableMappingsToParameterMappings(this.state.parameterMappings);

widget.options.parameterMappings = newMappings;

const somethingChanged = mapingsChanged(prevMappings, newMappings); 
// return `somethingChanged` from modal
```

Reply:
Not too excited about using `JSON.stringify` to comparison. While it works, it feels like a hack that can break in the future for unexpected reasons.

Maybe use [`_.isMatch`](https://lodash.com/docs/4.17.11#isMatch) instead?

(assuming we want to go this path)

Reply:
It's bad practice to do this:
```js
var a;
var b = process(a)
var diff = compare(a, b);
```
Since `process` has a different purpose, it might change one day and unintentionally mess the comparison up. (Unless we have a unit test for this of-course)

Reply:
In this case we can use [`_.some`](https://lodash.com/docs/4.17.11#some) to reduce amount of checks

Reply:
@ranbena You're not doing what is in you example. You do:
```javascript
var a;
var b = process(c); // <-- here
var diff = compare(a, b);
```
:wink:

Reply:
And once more: the only purpose of `editableMappingsToParameterMappings` is to create structure for `widget.options.parameterMappings`; it's the only function that creates that structure, it was intended for that. It has no side effects and by it's definition it returns new structure (and does not modify existing).

Reply:
I think it's a dangerous way to go about it, but it's not a big deal to me.
@kravets-levko, if you have a clear idea of how to implement this, I would appreciate if you could push a fix commit to this PR.

Reply:
> The object structure of `this.state.parameterMappings` and `widget.options.parameterMappings` is not the same.

@kravets-levko What I meant was that `editableMappingsToParameterMappings()` exposes only values of static params so it can't be used as input for comparison.